### PR TITLE
DOC: Add lib.format.open_memmap to autosummary.

### PR DESCRIPTION
**Pull request by rossbar**
An alternative to #13414 to address #12926.

Fixes link in the `See also` section of `np.memmap` docs and ensures
`lib.format.open_memmap` is included in the summary of io functions.

### CODE REVIEW

File: doc/source/reference/routines.io.rst [hunk 1]
```diff
@@ -85,7 +86,6 @@ Data sources
 Binary Format Description
 -------------------------
 .. autosummary::
-   :template: autosummary/minimal_module.rst
```
Comment (mattip) [inline doc/source/reference/routines.io.rst:88]: why this change? Is there something wrong with the template?

Comment (rossbar) [reply doc/source/reference/routines.io.rst:88]: It was based on [this comment](https://github.com/numpy/numpy/pull/13414#issuecomment-492908468) from the review of #13414. If no longer necessary I'm happy to revert.

Comment (mattip) [reply doc/source/reference/routines.io.rst:88]: Thanks, that overly-terse comment brought me back to [this one](https://github.com/numpy/numpy/pull/13414#issuecomment-487290004) which pointed out that `open_memmap` is [not actually documented](https://numpy.org/devdocs/search.html?q=open_memmap&check_keywords=yes&area=default), but after removing that line in this PR [it is](https://14195-908607-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.memmap.html) (scroll down to the See Also section). I think this is the only reference to that template, so maybe it can be removed. 

On the other hand, the [index page for memmap](https://14195-908607-gh.circle-artifacts.com/0/doc/build/html/reference/routines.io.html#memory-mapping-files) now seems weird: why does the function appear as `lib.format.open_memmap` and not `open_memmap`, and where are the rest of the functions from `lib.format.py` like `read_array`, `write_array` and the docstring for that file?

Comment (rossbar) [reply doc/source/reference/routines.io.rst:88]: > but after removing that line in this PR it is (scroll down to the See Also section). 

I think this is rather due to the fact that I explicitly added the function `lib.format.open_memmap` to the `routines.io` autosummary section on memmaps (see the other changed line in this PR).

> and where are the rest of the functions from lib.format.py like read_array, write_array and the docstring for that file?

I'm not sure - maybe the functions in `lib.format` are not being picked up by autosummary because the `lib.format.rst` doc is itself auto-generated?

Comment (rossbar) [reply doc/source/reference/routines.io.rst:88]: I tried several approaches but wasn't able to get the `autosummary` to include the docstrings for the `lib.format` functions in the generated docs. I wonder if they were ever included as it turns out `descr_to_dtype` had an rST formatting error in it which raised a sphinx warning on build.

In any case, eb6f013 presents an alternative which explicitly includes the `lib.format` functions in an autosummary in `routines.io` to ensure the `.rst` stubs for the functions are generated properly.

File: doc/source/reference/routines.io.rst [hunk 2]
```diff
@@ -85,7 +85,27 @@ Data sources
...
    :toctree: generated/
 
-    lib.format
+   lib.format
```
Comment (mattip) [inline doc/source/reference/routines.io.rst:90]: Now the function's documentation appears both in [lib.format](https://14213-908607-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.lib.format.html#module-numpy.lib.format) and on [this page](https://14213-908607-gh.circle-artifacts.com/0/doc/build/html/reference/routines.io.html), where in the current devdocs it [is not in lib.format](https://numpy.org/devdocs/reference/generated/numpy.lib.format.html#module-numpy.lib.format) nor on [this page](https://numpy.org/devdocs/reference/routines.io.html). Weird. Is that because of the change in indentation or because of the removal of the template? In any case do you want to leave the "Related Functions" section here as a duplication of the functions on the `lib.format` page?

Comment (rossbar) [reply doc/source/reference/routines.io.rst:91]: > Weird. Is that because of the change in indentation or because of the removal of the template?

This is because of the removal of the template. I've added the template back in in 1119338 to prevent the duplication.

> In any case do you want to leave the "Related Functions" section here as a duplication of the functions on the lib.format page?

1119338 Leaves the "Related functions" section but removes the function listing from the `lib.format` docstring by adding back the template. For whatever reason, the explicit listing of the functions in the `.. autosummary` in `routines.io` causes them all to be documented, whereas the autogenerated listing in `generated/numpy.lib.format.rst` doesn't. 

Now the listing of the functions in `numpy.lib.format` should appear only once (in the "Related functions" section of `routines.io`) and the html for each function's docstring is generated (and any links to `lib.format.<function>` should now work).

Comment (mattip) [reply doc/source/reference/routines.io.rst:91]: I would prefer to remove the template and the "Related Functions" section, and leave the semi-private functions documented on the `lib.format` page. Hopefully when we fill out the skeleton [how to read and write data](https://14217-908607-gh.circle-artifacts.com/0/doc/build/html/user/ionumpy.html) page the use of these functions will be clearer. 

> ... the explicit listing of the functions ... causes them all to be documented, whereas the autogenerated listing ... doesn't.

When I compare the ["Related Functions" listing](https://14217-908607-gh.circle-artifacts.com/0/doc/build/html/reference/routines.io.html#related-functions) to the [Functions in lib.format](https://14213-908607-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.lib.format.html#module-numpy.lib.format#Functions) page from eb6f013, they seem identical. What am I missing? 

In any case those functions need cleanup in follow-on PRs:
- the docstring of `descr_to_dtype` is confusing
- some of those seem like private functions

Comment (rossbar) [reply doc/source/reference/routines.io.rst:91]: > I would prefer to remove the template and the "Related Functions" section, and leave the semi-private functions documented on the lib.format page. Hopefully when we fill out the skeleton how to read and write data page the use of these functions will be clearer.

Sounds good - I agree, which is why I had originally only proposed adding the `lib.format.open_memmap` function to the `routines.io` to trigger the inclusion of its docstring in the html docs (and therefore make it linkable).

> When I compare the "Related Functions" listing to the Functions in lib.format page from eb6f013, they seem identical. What am I missing?

If you checkout the original commit (e7adb14), the listing of functions is included in the `lib.format` doc, but `.rst` stubs for the docstrings of each of the functions are not generated, so the individual function docstrings are not included in the html documentation (except for `lib.format.open_memmap`, which I added to one of the `.. autosummary`s in `routines.io`). The reason they are generated/linkable in eb6f013 is because each of the `lib.format` functions is manually included in the `autosummary` in the "Related functions" section of `routines.io`.

> In any case those functions need cleanup in follow-on PRs:
> * the docstring of descr_to_dtype is confusing

I opened #16673 to track this

> * some of those seem like private functions

Agreed, this is why I limited the changes originally to *only* including `lib.format.open_memmap`, as that was the function for which the `See Also` link in `memmap` was not linking.

Given the discussion, the thing that makes the most sense to me is to revert to e7adb14 which will:
 1. Add `lib.format.open_memmap` to the html documentation, advertise it publicly in `routines.io` in the refguide, and fix the link in the "See Also" section of `numpy.memmap`.
 2. Add the listing of all of the functions in `lib.format` to the bottom of the `lib.format` documentation, but *only* the docstring for `lib.format.open_memmap` will be generated. All the other functions will be summarized, but their docstrings will not be included in the html docs.

Does this seem like a good plan?

Comment (mattip) [reply doc/source/reference/routines.io.rst:91]: ok. It all seems like a mess to me so whatever you can do to move it forward without too much churn sounds good.

**Comment by mattip** [comment]
Thanks @rossbar. FWIW, now on the lib.format page the functions appear at the bottom with short docstrings.

**Comment by rossbar** [comment]
Thanks for the review @mattip !

> FWIW, now on the lib.format page the functions appear at the bottom with short docstrings.

Yes, but only `open_memmap` should be linkable (the other docstrings are still not [yet] autogenerated).
